File: src/test/java/no/nav/sosialhjelp/soknad/web/rest/ressurser/begrunnelse/BegrunnelseRessursTest.java
package no.nav.sosialhjelp.soknad.web.rest.ressurser.begrunnelse;

import no.nav.sbl.soknadsosialhjelp.soknad.begrunnelse.JsonBegrunnelse;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKildeBruker;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.model.exception.AuthorizationException;
import no.nav.sosialhjelp.soknad.domain.model.oidc.StaticSubjectHandlerService;
import no.nav.sosialhjelp.soknad.domain.model.oidc.SubjectHandler;
import no.nav.sosialhjelp.soknad.web.sikkerhet.Tilgangskontroll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static no.nav.sosialhjelp.soknad.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static no.nav.sosialhjelp.soknad.web.rest.ressurser.begrunnelse.BegrunnelseRessurs.BegrunnelseFrontend;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class BegrunnelseRessursTest {

    private static final String BEHANDLINGSID = "123";
    private static final String EIER = "123456789101";
    private static final String SOKER_FORDI = "Jeg søker fordi...";
    private static final String SOKER_OM = "Jeg søker om...";

    @Mock
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Mock
    private Tilgangskontroll tilgangskontroll;

    @InjectMocks
    private BegrunnelseRessurs begrunnelseRessurs;

    @BeforeEach
    public void setUp() {
        System.setProperty("environment.name", "test");
        SubjectHandler.setSubjectHandlerService(new StaticSubjectHandlerService());
    }

    @AfterEach
    public void tearDown() {
        SubjectHandler.resetOidcSubjectHandlerService();
        System.clearProperty("environment.name");
    }

    @Test
    void getBegrunnelseSkalReturnereBegrunnelseMedTommeStrenger(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithBegrunnelse("", ""));

        final BegrunnelseFrontend begrunnelseFrontend = begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID);

        assertThat(begrunnelseFrontend.hvaSokesOm).isBlank();
        assertThat(begrunnelseFrontend.hvorforSoke).isBlank();
    }

    @Test
    void getBegrunnelseSkalReturnereBegrunnelse(){
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithBegrunnelse(SOKER_OM, SOKER_FORDI));

        final BegrunnelseFrontend begrunnelseFrontend = begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID);

        assertThat(begrunnelseFrontend.hvaSokesOm).isEqualTo(SOKER_OM);
        assertThat(begrunnelseFrontend.hvorforSoke).isEqualTo(SOKER_FORDI);
    }

    @Test
    void putBegrunnelseSkalSetteBegrunnelse(){
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithBegrunnelse("", ""));

        final BegrunnelseFrontend begrunnelseFrontend = new BegrunnelseFrontend()
                .withHvaSokesOm(SOKER_OM)
                .withHvorforSoke(SOKER_FORDI);
        begrunnelseRessurs.updateBegrunnelse(BEHANDLINGSID, begrunnelseFrontend);

        final SoknadUnderArbeid soknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        final JsonBegrunnelse begrunnelse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getBegrunnelse();
        assertThat(begrunnelse.getKilde()).isEqualTo(JsonKildeBruker.BRUKER);
        assertThat(begrunnelse.getHvaSokesOm()).isEqualTo(SOKER_OM);
        assertThat(begrunnelse.getHvorforSoke()).isEqualTo(SOKER_FORDI);
    }

    @Test
    void getBegrunnelseSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerHarTilgang();

        assertThatExceptionOfType(AuthorizationException.class)
                .isThrownBy(() -> begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID));

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    @Test
    void putBegrunnelseSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(BEHANDLINGSID);

        var begrunnelseFrontend = new BegrunnelseFrontend().withHvaSokesOm(SOKER_OM).withHvorforSoke(SOKER_FORDI);

        assertThatExceptionOfType(AuthorizationException.class)
                .isThrownBy(() -> begrunnelseRessurs.updateBegrunnelse(BEHANDLINGSID, begrunnelseFrontend));

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    private SoknadUnderArbeid catchSoknadUnderArbeidSentToOppdaterSoknadsdata() {
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(soknadUnderArbeidRepository).oppdaterSoknadsdata(argument.capture(), anyString());
        return argument.getValue();
    }

    private SoknadUnderArbeid createJsonInternalSoknadWithBegrunnelse(String hvaSokesOm, String hvorforSoke) {
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getBegrunnelse()
                .withHvaSokesOm(hvaSokesOm)
                .withHvorforSoke(hvorforSoke);
        return soknadUnderArbeid;
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/begrunnelse/BegrunnelseRessursTest.kt
package no.nav.sosialhjelp.soknad.begrunnelse

import io.mockk.called
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.runs
import io.mockk.slot
import io.mockk.unmockkObject
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKildeBruker
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import no.nav.sosialhjelp.soknad.app.exceptions.AuthorizationException
import no.nav.sosialhjelp.soknad.app.subjecthandler.StaticSubjectHandlerImpl
import no.nav.sosialhjelp.soknad.app.subjecthandler.SubjectHandlerUtils
import no.nav.sosialhjelp.soknad.begrunnelse.BegrunnelseRessurs.BegrunnelseFrontend
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.SoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class BegrunnelseRessursTest {

    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository = mockk()
    private val tilgangskontroll: Tilgangskontroll = mockk()
    private val begrunnelseRessurs = BegrunnelseRessurs(tilgangskontroll, soknadUnderArbeidRepository)

    @BeforeEach
    fun setUp() {
        mockkObject(MiljoUtils)
        every { MiljoUtils.isNonProduction() } returns true
        SubjectHandlerUtils.setNewSubjectHandlerImpl(StaticSubjectHandlerImpl())
    }

    @AfterEach
    fun tearDown() {
        SubjectHandlerUtils.resetSubjectHandlerImpl()
        unmockkObject(MiljoUtils)
    }

    @Test
    fun begrunnelseSkalReturnereBegrunnelseMedTommeStrenger() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns createJsonInternalSoknadWithBegrunnelse("", "")
        val begrunnelseFrontend = begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID)
        assertThat(begrunnelseFrontend.hvaSokesOm).isBlank
        assertThat(begrunnelseFrontend.hvorforSoke).isBlank
    }

    @Test
    fun begrunnelseSkalReturnereBegrunnelse() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithBegrunnelse(SOKER_OM, SOKER_FORDI)

        val begrunnelseFrontend = begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID)
        assertThat(begrunnelseFrontend.hvaSokesOm).isEqualTo(SOKER_OM)
        assertThat(begrunnelseFrontend.hvorforSoke).isEqualTo(SOKER_FORDI)
    }

    @Test
    fun putBegrunnelseSkalSetteBegrunnelse() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns
            createJsonInternalSoknadWithBegrunnelse("", "")

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val begrunnelseFrontend = BegrunnelseFrontend(SOKER_OM, SOKER_FORDI)
        begrunnelseRessurs.updateBegrunnelse(BEHANDLINGSID, begrunnelseFrontend)

        val soknadUnderArbeid = soknadUnderArbeidSlot.captured
        val begrunnelse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.begrunnelse
        assertThat(begrunnelse.kilde).isEqualTo(JsonKildeBruker.BRUKER)
        assertThat(begrunnelse.hvaSokesOm).isEqualTo(SOKER_OM)
        assertThat(begrunnelse.hvorforSoke).isEqualTo(SOKER_FORDI)
    }

    @Test
    fun begrunnelseSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { begrunnelseRessurs.hentBegrunnelse(BEHANDLINGSID) }

        verify { soknadUnderArbeidRepository wasNot called }
    }

    @Test
    fun putBegrunnelseSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } throws AuthorizationException("Not for you my friend")

        val begrunnelseFrontend = BegrunnelseFrontend(SOKER_OM, SOKER_FORDI)
        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { begrunnelseRessurs.updateBegrunnelse(BEHANDLINGSID, begrunnelseFrontend) }

        verify { soknadUnderArbeidRepository wasNot called }
    }

    private fun createJsonInternalSoknadWithBegrunnelse(hvaSokesOm: String, hvorforSoke: String): SoknadUnderArbeid {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.begrunnelse
            .withHvaSokesOm(hvaSokesOm)
            .withHvorforSoke(hvorforSoke)
        return soknadUnderArbeid
    }

    companion object {
        private const val BEHANDLINGSID = "123"
        private const val EIER = "123456789101"
        private const val SOKER_FORDI = "Jeg søker fordi..."
        private const val SOKER_OM = "Jeg søker om..."

        private fun createSoknadUnderArbeid(): SoknadUnderArbeid {
            return SoknadUnderArbeid(
                versjon = 1L,
                behandlingsId = "behandlingsid",
                tilknyttetBehandlingsId = null,
                eier = EIER,
                jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
                status = SoknadUnderArbeidStatus.UNDER_ARBEID,
                opprettetDato = LocalDateTime.now(),
                sistEndretDato = LocalDateTime.now()
            )
        }
    }
}
